README.md
# nuvo-dashing-js

A fork of the port of [Dashing](http://shopify.github.io/dashing/) to node.js in attempt to revive the project with updated dependencies.

[![NPM version](https://badge.fury.io/js/nuvo-dashing-js.png)](http://badge.fury.io/js/nuvo-dashing-js)
[![Dependency Status](https://david-dm.org/lokulin/nuvo-dashing-js.png)](https://david-dm.org/lokulin/nuvo-dashing-js)

## Getting Started

1. Install from npm

```shell
$ npm install -g nuvo-dashing-js
```

2. Generate a new project

```shell
$ dashing-js new sweet_dashboard_project
```

3. Change your directory to sweet_dashboard_project and install required modules

```shell
$ cd sweet_dashboard_project && npm install
```

4. Start the server!

```shell
$ dashing-js start
```

5. Point your browser at http://localhost:3030/ and have fun!

***

Every new Dashing project comes with sample widgets & sample dashboards for you to explore. The directory is setup as follows:

* Assets — All your images, fonts, and js/coffeescript libraries. Uses <del>[Sprockets](https://github.com/sstephenson/sprockets)</del> [Mincer](http://nodeca.github.io/mincer/).
* Dashboards — One .jade file for each dashboard that contains the layout for the widgets.
* Jobs — Your js/coffee jobs for fetching data (e.g for calling third party APIs like twitter). Name them *\*.job.js/\*.job.coffee*
* Lib — Optional js/coffee files to help out your jobs.
* Public — Static files that you want to serve. A good place for a favicon or a custom 404 page.
* Widgets — All the html/css/coffee for individual widgets.

Run `dashing-js` from command line to find out what command line tools are available to you.

## Deployment
dashing-js is setup to be easily deployed on a Joyent Node SmartMachine. This means that:

1. The version of Node is defined in package.json
2. The main script to run is server.js
3. The web server port is pulled from process.env.PORT



HISTORY.md
0.1.11 / 2015.06.03
-------------------

* Refreshed dependencies

0.1.10 / 2015.05.03
-------------------

* Removed unused .js and .css files

0.1.9 / 2015.05.03
------------------

* Fix damage from dashing merge
* Re-update fontawesome to 4 series.


0.1.8 / 2015.05.02
------------------

* Close event socket after each message
* fixing: throwing error while loading file
* feat: import matching settings json file data into dashboard if it exists
* Re-pull changes from dashing


0.1.7 / 2015.05.02
------------------

* Fixed broken version 0.1.6


0.1.6 / 2015.05.02
------------------

* Updated dependecy versions


0.1.5 / 2013.12.27
-------------------

* Updated libraries versions
  - request (~2.29.0 -> ~2.30.0)
* Added IE9+ compatibility. Thanks to @arabold [PR #15]


0.1.4 / 2013.12.12
-------------------

* Fixed indentation in jade template to properly show up default widget. Thanks to @arabold [PR #13]


0.1.3 / 2013.12.09
-------------------

* Updated libraries versions
  - request (~2.27.0 -> ~2.29.0)
* Use different tags when scaffolding with ejs. Fixes #12
* Include custom stylesheets in ejs layout only if defined


0.1.2 / 2013.12.04
-------------------

* Handle widgets with multiple capital letters (original [PR #181](https://github.com/Shopify/dashing/pull/181))
* Use right arrow icon if difference is zero in number widget. Fixes #11
* Upgrade FontAwesome to version 4.0.3 and update new icon names ([Changes](https://github.com/FortAwesome/Font-Awesome/wiki/Upgrading-from-3.2.1-to-4))
* Use relative url for EventSource to allow running on sub-path (original [PR #209](https://github.com/Shopify/dashing/pull/209))
* Strip html by default in widgets. Users can disable this with the 'raw' filter. (orig [9f93895bd4](https://github.com/Shopify/dashing/commit/9f93895bd40aad02e88f7ed7bfd954c930aa27db))
* Update d3 (v3.3.11) and rickshaw (v1.4.5)


0.1.1 / 2013.12.03
-------------------

* Fix number widgets in sample dashboard
* Use right arrow icon if difference is zero in number widget. Fixes #11


0.1.0 / 2013.12.03
-------------------

* Updated libraries versions
  - commander (~1.1.1 -> ~2.1.0)
  - consolidate (~0.9.0 -> ~0.10.0)
  - express (~3.2.0 -> ~3.4.6)
  - fs-extra (~0.6.0 -> ~0.8.1)
  - jade (~0.29.0 -> ~0.35.0)
  - mincer (~0.4.5 -> ~0.5.12)
  - node-sass (0.5.2 -> ~0.7.0)
  - request (~2.20.0 -> ~2.27.0)
* Jade version 0.31.0 deprecated implicit text only support for scripts and styles.
* Added library prompt (commander removed confirm & prompt methods)
* Explicit flush response for SSE (connect 2.10.0)


0.0.10 / 2013.07.10
-------------------

* Fixed `node-sass` version to 0.5.2


0.0.9 / 2013.07.09
------------------

* Merge fidgety's PL to stick `node-sass` version to 0.5.2


0.0.8 / 2013.07.04
------------------

* Fixed stderr redirection to log file
* Fixed `winston` configuration
* Fixed cli support for job files in coffeescript


0.0.7 / 2013.06.12
------------------

* Refactor to use `winston` logging
* Fixed dameonize stdio redirection


0.0.6 / 2013.05.29
------------------

* Fixed support for coffee job files
* Fixed dameonize stdio redirection


0.0.5 / 2013.05.17
------------------

* Use js native milliseconds when updating data on widgets


0.0.4 / 2013.05.15
------------------

* Possibility to write jobs in coffeescript


0.0.3 / 2013.05.11
------------------

* Updated `node-sass` version, re-enabled widgets animations
* Updated FontAwesome assets to version 3.0.2
* Support for process daemonize


0.0.2 / 2013.05.02
------------------

* Added ability to install widgets from zipballs
* Added ability to add custom stylesheets
* Merge original [PR #115](https://github.com/Shopify/dashing/pull/115) (Added suffix option for number widget)


0.0.1 / 2013.04.26
------------------
* First release


lib/dashing.js
var fs = require('fs')
  , path = require('path')
  , express = require('express')
  , Mincer = require('mincer')
  , coffee = require('coffee-script')
  , morgan = require('morgan')
  , errorHandler = require('errorhandler')
  , compression = require('compression')
  , bodyParser = require('body-parser')
  , serveStatic = require('serve-static')
  , methodOverride = require('method-override')
  , InMemoryHistory = require('./inMemoryHistory')
  , Promise = require('bluebird');

module.exports.logger = logger = require('./logger');
module.exports.Dashing = function Dashing() {
  var dashing = {};
  dashing.root = path.resolve(__dirname, '../../..');
  dashing.NODE_ENV = process.env.NODE_ENV || 'development';

  dashing.view_engine = process.env.VIEW_ENGINE || 'jade';

  dashing.mincer = {};
  dashing.mincer.environment = new Mincer.Environment();
  dashing.mincer.assets_prefix = '/assets';
  dashing.mincer.environment.appendPath('assets/javascripts');
  dashing.mincer.environment.appendPath('assets/stylesheets');
  dashing.mincer.environment.appendPath('assets/fonts');
  dashing.mincer.environment.appendPath('assets/images');
  dashing.mincer.environment.appendPath('widgets');
  dashing.mincer.environment.appendPath(path.resolve(__dirname, '../javascripts'));

  dashing.public_folder = dashing.root + '/public';
  dashing.views = dashing.root + '/dashboards';
  dashing.settings = dashing.root + '/settings';
  dashing.default_dashboard = null;
  dashing.port = (process.env.PORT || 3030);
  dashing.force_close_event_socket = false;

  dashing.protected = function(req, res, next) {
    next();
  };

  dashing._protected = function(req, res, next) {
    dashing.protected(req, res, next);
  }

  var expressLoggerOptions = {
    format: 'dev',
    stream: {
      write: function(message, encoding) {
        logger.info(message);
      }
    }
  };

  // setup Express
  var app = express();
  if (app.get('env') == 'development') {
    Mincer.logger.use(logger);
  } else {
    expressLoggerOptions.format = 'short';
    // In production we assume that assets are not changed between requests,
    // so we use cached version of environment.
    // All file system methods are cached for the instances lifetime.
    dashing.mincer.environment = dashing.mincer.environment.index;
  }
  app.set('views', dashing.views);
  app.set('view engine', dashing.view_engine);
  if (dashing.view_engine === 'ejs') {
    app.use(require('express-ejs-layouts'));
  }
  app.use(morgan('dev'));
  app.use(errorHandler());
  app.use(compression());
  app.use(bodyParser.json());
  app.use(bodyParser.urlencoded());
  app.use(methodOverride());
  app.use(dashing.mincer.assets_prefix, Mincer.createServer(dashing.mincer.environment));
  app.use(serveStatic(dashing.public_folder));
  app.set('development', dashing.NODE_ENV === 'development');
  app.set('production', dashing.NODE_ENV === 'production');

  var connections = {};

  app.get('/events', dashing._protected, function(req, res) {
    // let request last as long as possible
    req.socket.setTimeout(0);

    if (!dashing.force_close_event_socket) {
      // this won't be necessary if we close the connection after each data package
      var conn = {
        id: (new Date().getTime().toString() + Math.floor(Math.random() * 1000).toString()),
        send: function(body) {
          res.write(body);
          res.flush(); // need to flush with .compress()
        }
      };
      connections[conn.id] = conn;
    }

    // send headers for event-stream connection
    res.writeHead(200, {
      'Content-Type': 'text/event-stream',
      'Cache-Control': 'no-cache',
      'Connection': 'keep-alive',
      'X-Accel-Buffering': 'no', // Disable buffering for nginx
      'Access-Control-Allow-Origin': '*' //fix XHR issue
    });
    res.write('\n');
    res.write(Array(2049).join(' ') + '\n'); // 2kb padding for IE
    latest_events()
      .then( function(events) {
        if(events) {
          res.write(events);
        }
      }).done(function() {
        res.flush(); // need to flush with .compress()
      });

    if (!dashing.force_close_event_socket) {
	    req.on('close', function() {
	      delete connections[conn.id];
	    });
    }
    else {
    	res.end();
    }
  });

  app.get('/', function(req, res) {
    if (dashing.default_dashboard) {
      res.redirect(dashing.default_dashboard);
    } else {
      first_dashboard(function(err, dashboard) {
        if (err) {
          next(err);
        } else if (dashboard) {
          res.redirect(dashboard);
        } else {
          next(new Error('There are no dashboards in your dashboard directory.'));
        }
      });
    }
  });

  app.get('/:dashboard', dashing._protected, function(req, res) {
    var dashboard = req.params.dashboard;
    fs.exists([dashing.views, dashboard + '.' + dashing.view_engine].join(path.sep), function(exists) {
      if (exists) {
        var templateVars = {
          dashboard: dashboard,
          request: req
        };
        var settingsPath = [dashing.settings, dashboard + '.json'].join(path.sep);
        fs.exists(settingsPath, function(exists) {
          if (exists) {
            templateVars.settings = require(settingsPath);
          }
          res.render(dashboard, templateVars);
        });
      } else {
        res.status(404).sendfile(dashing.public_folder + '/404.html');
      }
    });
  });

  app.get('/views/:widget?.html', dashing._protected, function(req, res) {
    var widget = req.params.widget;
    res.sendfile([dashing.root, 'widgets', widget, widget + '.html'].join(path.sep));
  });

  app.post('/widgets/:id', function(req, res) {
    var auth_token = req.body.auth_token;
    if (!dashing.auth_token || dashing.auth_token == auth_token) {
      send_event(req.params.id, req.body);
      res.send(204);
    } else {
      res.send(401, 'Invalid API key');
    }
  });

  // The 404 Route (ALWAYS Keep this as the last route)
  app.use(function(req, res, next) {
    res.status(404).sendfile(dashing.public_folder + '/404.html');
  });

  // Error handler
  app.use(function(err, req, res, next) {
    logger.error(err.stack);
    res.send(500, err);
  });

  function send_event(id, body) {
    body.id = id;
    body.updatedAt = Date.now();
    var event = format_event(body);
    dashing.history.record(id, event);
    for (var k in connections) {
      connections[k].send(event);
    }
  }
  global.send_event = send_event;

  function format_event(body) {
    return 'data: ' + JSON.stringify(body) + '\n\n';
  }

  function latest_events() {
    return Promise
      .resolve(dashing.history.getAll()) // assure trusted promise
      .reduce( function(agg, next) {
        return agg + next; 
      });
  }

  function first_dashboard(fn) {
    fs.readdir(dashing.views, function(err, files) {
      if (err) fn(err);
      var regex = new RegExp('(\w*)\.' + dashing.view_engine + '$');
      for (var i in files) {
        var file = files[i];
        if (file.match(regex) && file !== 'layout.' + dashing.view_engine) {
          fn(null, file.substring(0, file.length - (dashing.view_engine.length + 1)));
          return;
        }
      }
      fn(null, null);
    });
  }

  // Load custom libraries
  fs.readdir([dashing.root, 'lib'].join(path.sep), function(err, files) {
    if (err) throw err;
    for (var i in files) {
      var file = [dashing.root, 'lib', files[i]].join(path.sep);
      require(file);
    }
  });

  function start_jobs() {
    // Load jobs files
    var job_path = process.env.JOB_PATH || [dashing.root, 'jobs'].join(path.sep);
    logger.info('Searching for jobs in', job_path);
    fs.readdir(job_path, function(err, files) {
      if (err) throw err;
      for (var i in files) {
        var file = [job_path, files[i]].join(path.sep);
        if (file.match(/(\w*)\.job\.(js|coffee)$/)) {
          logger.info('Loading job file:', files[i]);
          try {
            require(file);
          } catch (e) {
            console.error('error while loading job (%s): ', file, e, e.stack);
          }
      }
    });
  }

  function start_express() {
    app.listen(dashing.port);
    logger.info('Listening on http://0.0.0.0:' + dashing.port + (process.env.__daemon === 'false' ? ', CTRL+C to stop' : ''));
  }

  dashing.useHistoryStore = function(store) {
    dashing.history = store;
  }

  dashing.start = function() {
    try {
      start_jobs();
      start_express();
    } catch(e) {
      logger.error("Unable to start dashing");
      logger.error(e.toString()); // winston is not catching these?
    }
  }
  dashing.useHistoryStore(new InMemoryHistory());
  dashing.app = app;
  return dashing;
};
